### PR TITLE
[WIP] Notify beacon for Debian/Ubuntu systems

### PR DESCRIPTION
**Pull request by ricardoasmarques**
### What does this PR do?

This PR implements an `aptnotify` that will trigger a "package refresh" on Uyuni when `apt-get` is used to manually install/update/remove packages (similar to the existing `zyppnotify` and `yumnotify`).

**TODO**

- [ ] Packaging: The `99aptnotify` file must be installed in `/etc/apt/apt.conf.d/99aptnotify`
- [ ] Packaging: The `aptnotify` file must be installed in `/usr/bin/aptnotify`

### What issues does this PR fix or reference?

- https://github.com/SUSE/spacewalk/issues/11561

### Previous Behavior
Remove this section if not relevant

### New Behavior
Remove this section if not relevant

### Tests written?

No

### Commits signed with GPG?

No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Comment by meaksh** [comment]
NOTE: As mentioned in the **TODO**, this new file needs to be deployed when installing the DEB packages. This requires we change the `debian.tar` in our OBS debian subprojects to make this file to be included.

We will be doing it ourselves at the time of merging this PR.

File: scripts/suse/apt/aptnotify [hunk 1]
```diff
@@ -0,0 +1,46 @@
+#!/usr/bin/python3
```
Comment (meaksh) [inline scripts/suse/apt/aptnotify:1]: Just wondering: is this python3 shebang causing any pain to a Python 2, in for example Ubuntu 16.04?

File: scripts/suse/apt/aptnotify [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+def _get_checksum():
+    """
+    Get the checksum of the Package Database.
+
+    Returns:
+        hexdigest
+    """
+    digest = hashlib.sha256()
+    with open(DPKG_PATH, "rb") as rpm_db_fh:
```
Comment (meaksh) [inline scripts/suse/apt/aptnotify:27]: Minor nitpick:

```suggestion
    with open(DPKG_PATH, "rb") as pkg_db_fh:
```

File: scripts/suse/apt/aptnotify [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        while True:
+            buff = rpm_db_fh.read(0x1000)
```
Comment (meaksh) [inline scripts/suse/apt/aptnotify:29]:
```suggestion
            buff = pkg_db_fh.read(0x1000)
```

File: scripts/suse/apt/aptnotify [hunk 4]
```diff
@@ -0,0 +1,46 @@
+#!/usr/bin/python3
+
+import os
+import hashlib
+
+CK_PATH = "/var/cache/salt/minion/rpmdb.cookie"
```
Comment (agraul) [inline scripts/suse/apt/aptnotify:6]: I think this should be 
```suggestion
CK_PATH = "/var/cache/salt/minion/dpkg.cookie"
```

File: scripts/suse/apt/aptnotify [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+                break
+            digest.update(buff)
+    return digest.hexdigest()
+
+
+def dpkg_post_invoke():
+    """
+    Hook after the package installation transaction.
+    """
+    if 'SALT_RUNNING' not in os.environ:
```
Comment (agraul) [inline scripts/suse/apt/aptnotify:40]: We need to make sure that `SALT_RUNNING` is set for this to work

Comment (meaksh) [reply scripts/suse/apt/aptnotify:40]: Hmmm, not in this case. Here we're in the APT plugin, which always run when there is apt transaction.

In this case, we want to avoid changing the cookie if the `SALT_RUNNING` variable is set, and it's only been set by Salt aptpkg module when calling apt.

Changing the cookie will trigger the beacon event and therefore a pkg list refresh, and we don't want that when the apt execution was made by Salt.

File: scripts/suse/apt/aptnotify (hunk 4 above)
Comment (meaksh) [reply scripts/suse/apt/aptnotify:6]: You're right, this is a good point.

I think they decided to not changing this filename at this point because that would require refactors on the Java backend that generated the pkgset beacon configuration.

Currently, the configuration for the pkgset beacon is always defining `/var/cache/salt/minion/rpmdb.cookie` as the source for the cookie, regardless the OS.

File: scripts/suse/apt/aptnotify (hunk 5 above)
Comment (agraul) [reply scripts/suse/apt/aptnotify:40]: My comment was too easy to misunderstand. Let me try again ;)

We need to make sure that aptpkg is setting `SALT_RUNNING`, just like yumpkg and zypperpkg do. aptpkg uses the same mechanism that zypperpkg does, i.e. setting `SALT_RUNNING` can be done in the configuration file. I didn't check if it is already configured to do so, but I think not (given the bug report we had because we didn't configure it for zypperpkg). Either way, we should not forget this step.

File: scripts/suse/apt/aptnotify (hunk 4 above)
Comment (agraul) [reply scripts/suse/apt/aptnotify:6]: What about a generic name in that case? Could be named after the virtual pkg module: "pkg.cookie". Looking at the [Java code](https://github.com/uyuni-project/uyuni/blob/master/java/code/src/com/suse/manager/webui/services/pillar/MinionGeneralPillarGenerator.java#L47), the name should be easy to change.

Comment (rjmateus) [reply scripts/suse/apt/aptnotify:6]: Hey, I agree we should have a generic name but I'm not sure what will happen after we change it.
The problem would be for existing costumer with everything already in place. We would force them to upgrade all the clients tools to generate the cookie file with the new name or beacon will not work after change the java code side.
Do we really want to figure out what will be all the side effects of changing it?

Comment (isbm) [reply scripts/suse/apt/aptnotify:6]: Nothing special will happen if you change it. Just make sure you update minion *and* the backend.

Comment (mcalmer) [reply scripts/suse/apt/aptnotify:6]: But we cannot guarantee that the custom update both at the same time. I am not sure what we do exactly on the java side, but maybe we can do it for both names, old and new. After some time we can remove the old name and only leave the generic name.

Comment (paususe) [reply scripts/suse/apt/aptnotify:6]: Stupid proposal: symlink :-)

hey @isbm , glad to see you around!

**Comment by paususe** [comment]
I know it's my fault but shouldn't this be called `dpkgnotify` instead of `aptnotify`? After all, it should run after dpkg, since you may install/uninstall/update a package manually with dpkg (same as you can with rpm).

**Comment by meaksh** [comment]
I think this can be closed in favor of https://github.com/openSUSE/salt/pull/347 (which is already approved). I'm closing this draft PR now.

Feel free to reopen if necessary. Thanks!
